client/db/bolt/db.go

```diff
@@ -1514,31 +1556,93 @@ func (db *BoltDB) withBucket(bkt []byte, viewer txFunc, f bucketFunc) error {
 	})
 }
 
-// Backup makes a copy of the database.
-func (db *BoltDB) Backup() error {
-	return db.backup("")
+func ovrFlag(overwrite bool) int {
+	if overwrite {
+		return os.O_TRUNC
+	}
+	return os.O_EXCL
 }
 
-// backup makes a copy of the database to the specified file name in the backup
-// subfolder of the current DB file's folder. If fileName is empty, the current
-// DB's file name is used.
-func (db *BoltDB) backup(fileName string) error {
-	dir := filepath.Join(filepath.Dir(db.Path()), backupDir)
-	if _, err := os.Stat(dir); os.IsNotExist(err) {
-		err := os.Mkdir(dir, 0700)
+// compact writes a compacted copy of the DB into the specified destination
+// file. This function should be called from BackupTo to validate the
+// destination file and create any needed parent folders.
+func (db *BoltDB) compact(dst string, overwrite bool) error {
+	opts := bbolt.Options{
+		OpenFile: func(path string, _ int, mode os.FileMode) (*os.File, error) {
+			return os.OpenFile(path, os.O_RDWR|os.O_CREATE|ovrFlag(overwrite), mode)
+		},
+	}
+	newDB, err := bbolt.Open(dst, 0600, &opts)
+	if err != nil {
+		return fmt.Errorf("unable to compact database: %w", err)
+	}
+
+	const txMaxSize = 1 << 19 // 512 KiB
+	err = bbolt.Compact(newDB, db.DB, txMaxSize)
+	if err != nil {
+		_ = newDB.Close()
+		return fmt.Errorf("unable to compact database: %w", err)
+	}
+	return newDB.Close()
+}
+
+// compact writes a direct copy of the DB into the specified destination file.
+// This function should be called from BackupTo to validate the destination file
+// and create any needed parent folders.
+func (db *BoltDB) backup(dst string, overwrite bool) error {
+	// Just copy. This is like tx.CopyFile but with the overwrite flag set
+	// as specified.
+	f, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|ovrFlag(overwrite), 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	err = db.View(func(tx *bbolt.Tx) error {
+		_, err = tx.WriteTo(f)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	return f.Sync()
+}
+
+// BackupTo makes a copy of the database to the specified file, optionally
+// overwriting and compacting the DB.
+func (db *BoltDB) BackupTo(dst string, overwrite, compact bool) error {
+	// If relative path, use current db path.
+	var dir string
+	if !filepath.IsAbs(dst) {
+		dir = filepath.Dir(db.Path())
+		dst = filepath.Join(dir, dst)
+	}
+	dst = filepath.Clean(dst)
+	dir = filepath.Dir(dst)
+	if dst == filepath.Clean(db.Path()) {
+		return errors.New("destination is the active DB")
+	}
```

Comment:
My view of `dex.CleanAndExpandPath` is that it should be used only at the highest levels, such as in the cli apps, rpc and web server handlers, etc. rather than internal to these lower level packages.